### ISSUE TITLE
BUG: PyArray_MultiIter_SIZE is not defined in 1.26.4 headers despite documentation stating it is new in 1.26.0

### ISSUE THREAD

**Issue by oleksandr-pavlyk** [numpy/numpy#26990]
### Describe the issue:

Doc page https://numpy.org/doc/stable/reference/c-api/array.html for 2.0 stable states that `PyArray_MultiIter_SIZE` is new in 1.26.0.

It is defined with NumPy 2.0.

Lack of this function makes it difficult to update `mkl_random` to NumPy 2.0

### Reproduce the code example:

```python
(b_mkl_packages) opavlyk@opavlyk-mobl:~/repos/mkl_random$ conda list numpy
# packages in environment at /home/opavlyk/mamba3/envs/b_mkl_packages:
#
# Name                    Version                   Build  Channel
numpy                     1.26.4                   pypi_0    pypi
(b_mkl_packages) opavlyk@opavlyk-mobl:~/repos/mkl_random$ grep -r PyArray_MultiIter_SIZE $CONDA_PREFIX/lib/python3.12/site-packages/numpy
(b_mkl_packages) opavlyk@opavlyk-mobl:~/repos/mkl_random$
```
```


### Error message:

```shell
Building native extension in `mkl_random` fails with the like of


mkl_random/mklrand.cpp:10517:26: error: ‘PyArray_MultiIter_SIZE’ was not declared in this scope; did you mean ‘PyArray_MultiIter_GOTO’?
10517 |             __pyx_t_10 = PyArray_MultiIter_SIZE(__pyx_v_multi);
      |                          ^~~~~~~~~~~~~~~~~~~~~~
      |                          PyArray_MultiIter_GOTO
mkl_random/mklrand.cpp:10732:18: error: ‘PyArray_MultiIter_SIZE’ was not declared in this scope; did you mean ‘PyArray_MultiIter_GOTO’?
10732 |     __pyx_t_1 = (PyArray_MultiIter_SIZE(__pyx_v_multi) != __pyx_v_res_size);
      |                  ^~~~~~~~~~~~~~~~~~~~~~
      |                  PyArray_MultiIter_GOTO
mkl_random/mklrand.cpp:10970:24: error: ‘PyArray_MultiIter_NDIM’ was not declared in this scope; did you mean ‘PyArray_MultiIter_NEXT’?
10970 |     __pyx_v_multi_nd = PyArray_MultiIter_NDIM(__pyx_v_multi);
      |                        ^~~~~~~~~~~~~~~~~~~~~~
      |                        PyArray_MultiIter_NEXT
mkl_random/mklrand.cpp:10979:26: error: ‘PyArray_MultiIter_DIMS’ was not declared in this scope; did you mean ‘PyArray_MultiIter_DATA’?
10979 |     __pyx_v_multi_dims = PyArray_MultiIter_DIMS(__pyx_v_multi);
      |                          ^~~~~~~~~~~~~~~~~~~~~~
      |                          PyArray_MultiIter_DATA
```
```


### Python and NumPy Versions:

Python 3.12, NumPy 1.26.4 

### Runtime Environment:

_No response_

### Context for the issue:

_No response_

**Comment by oleksandr-pavlyk** [comment]
![image](https://github.com/user-attachments/assets/fdd87120-b8af-44be-9772-9cf5e45048ef)

**Comment by ngoldbaum** [comment]
It seems @charris found it too hard to backport and it never shipped: https://github.com/numpy/numpy/pull/24658#issuecomment-1883549864

I don't think we're doing updates to the 1.26 branch anymore so I think this might just be a known issue in a past release of NumPy?

**Comment by ngoldbaum** [comment]
Ping @seberg since you might have ideas to ease the C API transition for `mkl_random`.

**Comment by seberg** [comment]
Either force compiling with 2.0 or copy a custom definintion of the struct over (e.g. from the Cython 2 `.pyd`.

**Comment by oleksandr-pavlyk** [comment]
This is what I ended up doing https://github.com/IntelPython/mkl_random/blob/792b3f7181ef2970a71da26a1b1a0164926fd7cc/mkl_random/src/numpy_multiiter_workaround.h#L46-L78

**Comment by seberg** [comment]
Going to close.  Might have been nice to backport, but it was largely a Cython 2->3 thing, IIRC, and didn't seem used a lot even then.

The workarounds are not that bad (otherwise I am sure we would have backported it).

**Comment by seberg** [comment]
@oleksandr-pavlyk one thing I just realized, I think your error above is a fluke.  If you are *using* these macro's the Cython `.pyx` was generated with NumPy 2.0.

That is a problem I have seen before:  Cython doesn't notice that regeneration is required due to a dependency `.pyd` change (and those files can be hidden deep in the project and/or build folders).

The problem that *does* exist is that if you compile with Cython 3 (which you should), you cannot access the fields of them.

**TL;DR: Unless you are using the `broadcast` attributes in your code, this problem is a build issue and can be fixed by thoroughly cleaning up the generated `.c`/`.cpp` files.**

**Comment by WarrenWeckesser** [comment]
I just ran into this in an enhancement that I am working on for SciPy, written in Cython: https://github.com/scipy/scipy/pull/24011.

I'm experimenting with something like @oleksandr-pavlyk's `numpy_multiiter_workaround.h` that I can use in Cython to get the module to build with 1.26.  Any suggestions would be welcome.

**Comment by seberg** [comment]
Not sure it is the same issue, but the `__init__.pyd` didn't expose these macros at all until NumPy 2.0, I think?  So if you want to use those, you'll have to require 2.0 for the build.

Dunno about the CI setup, but I think that is OK these days.  Otherwise, I guess you'll have to copy-paste over the definitions from the `__init__.pyd` file.

(Dunno if you may need that custom header hack, although one could just embed it into the cython code if you want to compile with NumPy 1.26 for some reason.)

**Comment by WarrenWeckesser** [comment]
SciPy's minimum required version for NumPy is still 1.26, and that includes using it to build SciPy.  We have a few CI jobs that build with 1.26.

> (Dunno if you may need that custom header hack, although one could just embed it into the cython code if you want to compile with NumPy 1.26 for some reason.)

Thanks, embedding the compatibility shims in the Cython file is working fine.

**Comment by seberg** [comment]
> SciPy's minimum required version for NumPy is still 1.26, and that includes using it to build SciPy. We have a few CI jobs that build with 1.26.

Yeah, but that doens't mean you need *build* SciPy with it, the build could just use NumPy 2, you just need to run the tests with it.
